ckanext/security/anti_csrf.py

```diff
@@ -73,45 +84,155 @@ def get_cookie_token(request):
+    message_parts = message.split('/', 2)
+
+    return {
+        "message": message,
+        "hash": unicode(parts[0]),
+        "timestamp": int(message_parts[0]),
+        "nonce": int(message_parts[1]),
+        "username": message_parts[2]
+    }
+
+
 def get_response_token(request, response):
     """Retrieve the token to be injected into pages.
 
-    This will be retrieved from the 'token' cookie, if it exists and is fresh.
+    This will be retrieved from the 'token' cookie, if it exists and is valid.
     If not, a new token will be generated and a new cookie set.
     """
     # ensure that the same token is used when a page is assembled from pieces
-    if TOKEN_FIELD_NAME in request.cookies and TOKEN_FRESHNESS_COOKIE_NAME in request.cookies:
+    if TOKEN_FIELD_NAME in request.cookies:
         LOG.debug("Obtaining token from cookie")
         token = request.cookies.get(TOKEN_FIELD_NAME)
-        if not HEX_PATTERN.match(token):
-            LOG.debug("Invalid cookie token; making new token cookie")
+        if not validate_token(token) or is_soft_expired(token):
+            LOG.debug("Invalid or expired cookie token; making new token cookie")
             token = create_response_token(response)
     else:
-        LOG.debug("No fresh token found; making new token cookie")
+        LOG.debug("No valid token found; making new token cookie")
         token = create_response_token(response)
     return token
 
 
+def is_soft_expired(token):
+    """Check whether the token is old enough to need rotation.
+    It may still be valid, but it's time to generate a new one.
+
+    The current rotation age is 10 minutes.
+    TODO Make soft-rotation age configurable.
+    """
+    if not validate_token(token):
+        return False
+
+    now = int(time.time())
+    token_values = read_token_values(token)
+
+    return now - token_values['timestamp'] > 60 * 10
+
+
+def _get_secret_key():
+    """ Retrieve the secret key to use in generating secure hashes.
+    Currently this is the Beaker session secret.
+    """
+    return config.get('beaker.session.secret')
+
+
+def get_digest(message):
+    """ Generate a secure (unforgeable) hash of the provided data.
+    """
+    return hmac.HMAC(_get_secret_key(), message, hashlib.sha512).hexdigest()
+
+
 def create_response_token(response):
-    import binascii, os
-    token = binascii.hexlify(os.urandom(32))
-    response.set_cookie(TOKEN_FIELD_NAME, token, secure=True, httponly=True)
-    response.set_cookie(TOKEN_FRESHNESS_COOKIE_NAME, '1', max_age=600, secure=True, httponly=True)
+    """ Generate an unforgeable CSRF token. The format of this token is:
+    hash!timestamp/nonce/username
+    where the hash is a secure HMAC of the other values plus a secret key.
+    """
+    username = _get_safe_username()
+    timestamp = int(time.time())
+    nonce = random.randint(1, 999999)
+    message = "{}/{}/{}".format(timestamp, nonce, username)
```

Comment:
I guess the message ends with a `/` when the user is not logged in? I don't see any real issues with that at present, just a note. If there's no username at the end it should still validate as it'll compare two empty strings I think.

Reply:
Actually, it will fail if not logged in; the pattern used to parse the token doesn't allow blank usernames. I suppose that means it can't currently defend against Login CSRF. I'll take a look at enhancing it for that.

Reply:
FYI The latest `develop` changes to ckanext-csrf-filter are able to address Login CSRF, although blocking it requires an additional configuration step, to change the form plugin in `who.ini`.